cmd/tendermint/commands/debug/io.go

```diff
@@ -0,0 +1,117 @@
+package debug
+
+import (
+	"archive/zip"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// zipDir zips all the contents found in src, including both files and
+// directories, into a destination file dest. It returns an error upon failure.
+// It assumes src is a directory.
+func zipDir(src, dest string) error {
+	zipFile, err := os.Create(dest)
+	if err != nil {
+		return err
+	}
+	defer zipFile.Close()
+
+	zipWriter := zip.NewWriter(zipFile)
+	defer zipWriter.Close()
+
+	dirName := filepath.Base(dest)
+	baseDir := strings.TrimSuffix(dirName, filepath.Ext(dirName))
+
+	filepath.Walk(src, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		header, err := zip.FileInfoHeader(info)
+		if err != nil {
+			return err
+		}
+
+		// Each execution of this utility on a Tendermint process will result in a
+		// unique file.
+		header.Name = filepath.Join(baseDir, strings.TrimPrefix(path, src))
+
+		// Handle cases where the content to be zipped is a file or a directory,
+		// where a directory must have a '/' suffix.
+		if info.IsDir() {
+			header.Name += "/"
+		} else {
+			header.Method = zip.Deflate
+		}
+
+		headerWriter, err := zipWriter.CreateHeader(header)
+		if err != nil {
+			return err
+		}
+
+		if info.IsDir() {
+			return nil
+		}
+
+		file, err := os.Open(path)
+		if err != nil {
+			return err
+		}
+		defer file.Close()
+
+		_, err = io.Copy(headerWriter, file)
+		return err
+	})
+
+	return nil
+}
+
+// copyFile copies a file from src to dest and returns an error upon failure. The
+// copied file retains the source file's permissions.
+func copyFile(src, dest string) error {
+	if _, err := os.Stat(src); err != nil {
```

Comment:
why call Stat if you don't need statistics?

Reply:
This is the idiomatic way to check if the file exists before attempting to use/open it. I've added an explicit `os.IsNotExist(err)` check to make that clear.